tests/optimisation_test/recipe_converter.py

```diff
@@ -3,58 +3,84 @@
-                    _values = _output_pd[_head].tolist()
-                    print(_key,_values)
-                    _data_update = nested_update(self.data, key = str(_key), value = _values)
-                    self.data = _data_update
-                elif _key != _head:
-                    pass      
-
-    def create_optimized_recipe(self, path):
+            for col in range(len(param_dict)):
+                if _key == _keys_optimised_parameters_dict[col]:
+                    if _type == 'all':
+                        _value = list(param_dict[_keys_optimised_parameters_dict[col]].values())
+                    elif _type == 'any':
+                        pass
+                        _value = float(param_dict[_keys_optimised_parameters_dict[col]][0])
+                    else:
+                        logging.error('improper placeholder type')
+                    print(_value)
+                    _data_update = nested_update(recipe_dict, key = str(_key), value = _value)
+                    recipe_dict = _data_update
+                elif _key != _keys_optimised_parameters_dict[col]:
+                    pass
+        return recipe_dict    
+
+    def generate_recipe(self, _recipe_dict, path):
         with open(path, 'w') as file:
-            print(self.data)
-            yaml.dump(self.data, file)
+            print(_recipe_dict)
+            yaml.dump(_recipe_dict, file)
 
 if __name__ == "__main__": 
-    path = Path.cwd() / "tests" / "optimisation_test" / "algae_bot_recipe_template.yaml"
-    optimised_values = {'water': [29.0, 32.0, 34.0, 37.0, 40.0, 43.0], 'dye_A': [0.3, 0.33, 0.35, 0.38, 0.41, 0.44], 'dye_B': [0.31, 0.34, 0.36, 0.39, 0.42, 0.45]}
-    _pd = pd.DataFrame(optimised_values)
-    new_file_name = 'algae_bot_recipe'
-    data = YamlHandler.load_recipe_file(path)
+    cwd_path = Path.cwd()
+    recipe_path = Path.joinpath(cwd_path, "tests/optimisation_test/algae_bot_recipe_template.yaml")
+    optimised_parameters = {'water': [29.0, 32.0, 34.0, 37.0, 40.0, 43.0], 'dye_A': [0.3, 0.33, 0.35, 0.38, 0.41, 0.44], 'dye_B': [0.31, 0.34, 0.36, 0.39, 0.42, 0.45]}
+    _pd = pd.DataFrame(optimised_parameters)
+    optimised_parameters_dict = _pd.to_dict()
+    recipe_dict = YamlHandler.load_recipe_file(recipe_path)
+    print(recipe_dict)
     recipe = RecipeUpdate()
-    recipe.find_placeholders(data)
-    recipe.update_values(data, _pd)
-    recipe.create_optimized_recipe(path)
-    optimised_parameters = _pd.to_dict()
+    _keys_values, _keys_lists = recipe.find_placeholders(recipe_dict)
+    if len(_keys_values) != 0:
+        recipe_dict = recipe.update_values(recipe_dict,'any', optimised_parameters_dict)
+    else:
+        print('{{float.any}} place holders NOT found')
+
+    if len(_keys_lists) != 0:
+        recipe_dict = recipe.update_values(recipe_dict, 'all', optimised_parameters_dict)
+    else:
+        print('{{float.all}} place holders NOT found')    
+    recipe.generate_recipe(recipe_dict, recipe_path)
```

Comment:
Check comments above. I suggest making all of that internal methods and follow such execution sequence in main:
```python
...
recipe_dict = YamlHandler.load_recipe_file(recipe_path)
print(recipe_dict)
recipe_converter = RecipeUpdate(path_to_template_recipe, path_to_recipes_dir)
recipe_converter.generate_recipe(optimised_parameters_dict) # alternatively this can take _pd as input and convert to dict internally
```